Add `InverseOp` for `IsNonssocWord`

TWiedemann opened the pull request:
A nonassociative word (i.e., an element of a free magma or of a free magma-with-one) is invertible if and only if it equals the identity element (if such an element exists).

This is a draft because not all the tests that I added pass:

```
gap> M := FreeMagmaWithOne(2);
<free magma-with-one on the generators [ x1, x2 ]>
gap> IsMultiplicativeElementWithInverse(One(M));
false
```
We should have `IsMultiplicativeElementWithInverse(One(M)) = true`. I think the filter `IsNonassocWordWithOne` should imply `IsMultiplicativeElementWithInverse`, but I do not know how to implement that. (Note that `IsMultiplicativeElementWithInverse(x) = true` only implies that `x` is an admissible argument for `Inverse`, but it is allowed that `Inverse(x) = fail`.) If this is done, the PR should be complete.

@ThomasBreuer is listed as an author in the file.

## Text for release notes

none.

ThomasBreuer commented:
If we want to have an implication from `IsNonassocWordWithOne` to `IsMultiplicativeElementWithInverse` then the easiest way to achieve this is to change the declaration of `IsNonassocWordWithOne` from
```
DeclareSynonym( "IsNonassocWordWithOne", IsNonassocWord and IsWordWithOne );
```
to
```
DeclareSynonym( "IsNonassocWordWithOne", IsNonassocWord and IsWordWithInverse );
```
The question is if we want this.
For example, also the identity element of a free monoid is not regarded as invertible, since inverting is not a supported operation for monoids.

TWiedemann commented:
> The question is if we want this. For example, also the identity element of a free monoid is not regarded as invertible, since inverting is not a supported operation for monoids.

My impression is that there is nothing wrong with having an `Inverse` operation which may return `fail`. Even more so if it only returns `fail` for elements which are known to be non-invertible (as is the case in my suggestions for free magmas, and could be implemented similarly in free monoids). For example, currently `IsMultiplicativeElementWithInverse(Zero(Integers)) = true` and `Inverse(Zero(Integers)) = fail`. As another example,
```
gap> G := SymmetricGroup(3);
Sym( [ 1 .. 3 ] )
gap> F := FreeMagmaRing(Integers, G);
<free left module over Integers, and ring-with-one, with 2 generators>
gap> Inverse(F.1 + F.2);
fail
```

## Context

(Just to explain why I am so keen on inverting trivial elements.)

The reason I am interested in having `Inverse(One(M)) = One(M)` for a free magma `M` is that I am working with `F := FreeMagmaRing(K, M)` for a field `K` of rational functions. I would like to have `Inverse(x*One(F)) = x^-1 * One(F)` for `x in K`. The code for magma rings in GAP already supports `Inverse(x*m)` this for elements `m` in the magma `M` for which `Inverse` is defined and not `fail` (`mgmring:450-463`).

The reason I am interested in inverting elements of the form `x*One(F)` is that I want to evaluate certain rational functions `f in K` on elements of `F`, i.e., `Value(f, indetList, valueList, One(F))` where `indetList` is a list of indeterminates in the rational function field `K` and `valueList` is a list of elements of `F`. Clearly this does not work for arbitrary rational functions `f`, but I only want to use it for functions `f` whose denominator does not contain indeterminates from `indetList`, which implies that the denominator evaluates to an element of the form `x*One(F)` for some `x in K`.

With my suggested PR, all of the above works without any additional changes in rational functions or magma rings.

fingolfin commented:
Various test failures:
```
  ########> Diff in /home/runner/work/gap/gap/tst/testinstall/mgmfree.tst:279
  # Input is:
  One(M)^-1 = One(M);
  # Expected output:
  true
  # But found:
  Error, ^ cannot be used here to compute roots (use `RootInt' instead?)
  ########
  ########> Diff in /home/runner/work/gap/gap/tst/testinstall/mgmfree.tst:281
  # Input is:
  IsMultiplicativeElementWithInverse(M.1);
  # Expected output:
  true
  # But found:
  false
  ########
  ########> Diff in /home/runner/work/gap/gap/tst/testinstall/mgmfree.tst:283
  # Input is:
  IsMultiplicativeElementWithInverse(One(M));
  # Expected output:
  true
  # But found:
  false
  ########
```

TWiedemann commented:
@ThomasBreuer Thanks for having a detailed look! I made the following changes in the recent push:
1. Changed `DeclareSynonym( "IsNonassocWordWithOne", IsNonassocWord and IsWordWithOne );` to `DeclareSynonym( "IsNonassocWordWithOne", IsNonassocWord and IsWordWithInverse );`, as you suggested.
2. As a consequence of 1., `\^` is automatically defined for elements of a free magma-with-one and for negative integers by the method for `IsMultiplicativeElementWithInverse` in arith.gi. However, this method does not apply for elements of a free magma (without one), for which the proposed `InverseOp` also applies (because it is defined for `IsNonassocWord`). Hence I added your proposed method (with the additional check `IsMultiplicativeElementWithOne`), which I assume is also a tiny bit faster because it does not compute powers of one.
3. I moved `InverseOp` for `IsNonassocWord` to the file word.gi (from mgmfree.gi), and I also put `\^` there. It is my impression that word.gi is supposed to contain methods on the elements while mgmfree.gi contains methods on the free magma itself, but I am not sure.

As a consequence, `Inverse(x)` and `x^-1` are defined for all `x` in a free magma or in a free magma-with-one, and they are distinct from `fail` if and only if `x` is the identity element. Further, any `x` in a free magma-with-one has `IsMultiplicativeElementWithInverse(x) = true`, as it should because `One` and `Inverse` are defined for `x`.

I also have some thoughts on the associative case, but I think it is better to close the nonassociative case by merging this PR (if everything is ok) and to have a look at the associative case in a separate issue.

TWiedemann commented:
Comment: For consistency, I think it would be nice to define a similar function `InverseOp` for elements of free semigroups and free monoids. In fact, it would be sufficient to replace the requirement `IsNonassocWord` in the method in this PR by `IsWord`. However, this would imply that elements of free monoids should satisfy `IsMultiplicativeElementWithInverse` and hence, by the definitions in #5994, `IsAssocWordWithInverse`, which is currently a synonym for `IsElementOfFreeGroup`. Hence it seems dangerous to meddle with this.

ThomasBreuer reviewed (approved):
The underlying use case looks sensible. A "clean" solution by changing certain filters is not so easy, due to the fact that the current situation is not consistent. We can discuss this further, but I think for the moment we can just merge this pull request.

fingolfin commented on lib/word.gi:
```diff
@@ -269,3 +269,33 @@ InstallMethod( OneOp,
+    "for free magma element",
+    [ IsNonassocWord ],
+    function( x )
+    if IsMultiplicativeElementWithOne( x ) and x = One( x ) then
```
How about
```suggestion
    if IsMultiplicativeElementWithOne( x ) and IsOne( x ) then
```

fingolfin commented:
I've set the label indicating that this should *not* be mentioned in the release notes. Do you agree, @ThomasBreuer ? I am also fine with mentioning it but then someone (who is not me) should write a text for that ;-)
